File: ts/test/entity/combined_search_with_field/CombinedSearchWithFieldEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { PoetrydbSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('CombinedSearchWithFieldEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when POETRYDB_TEST_LIVE=TRUE.
  afterEach(liveDelay('POETRYDB_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = PoetrydbSDK.test()
    const ent = testsdk.CombinedSearchWithField()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.POETRYDB_TEST_LIVE
    for (const op of ['list']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'combined_search_with_field.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[],"name":"combined_search_with_field","op":{"list":{"input":"data","name":"list","points":[{"active":true,"args":{"params":[{"active":true,"example":"author","kind":"param","name":"input_field1","orig":"input_field1","reqd":true,"type":"`$STRING`","index$":0},{"active":true,"example":"linecount","kind":"param","name":"input_field2","orig":"input_field2","reqd":true,"type":"`$STRING`","index$":1},{"active":true,"example":"lines","kind":"param","name":"output_field","orig":"output_field","reqd":true,"type":"`$STRING`","index$":2},{"active":true,"example":"Shakespeare","kind":"param","name":"search_term1","orig":"search_term1","reqd":true,"type":"`$STRING`","index$":3},{"active":true,"example":"14","kind":"param","name":"search_term2","orig":"search_term2","reqd":true,"type":"`$STRING`","index$":4}]},"contract":{"id":"GET /{inputField1},{inputField2}/{searchTerm1};{searchTerm2}/{outputFields}","json":"{\"operationId\":\"combinedSearchWithFields\",\"parameters\":[{\"description\":\"First input field (author, title, lines, linecount)\",\"example\":\"author\",\"in\":\"path\",\"name\":\"inputField1\",\"required\":true,\"schema\":{\"enum\":[\"author\",\"title\",\"lines\",\"linecount\"],\"type\":\"string\"}},{\"description\":\"Second input field (author, title, lines, linecount)\",\"example\":\"linecount\",\"in\":\"path\",\"name\":\"inputField2\",\"required\":true,\"schema\":{\"enum\":[\"author\",\"title\",\"lines\",\"linecount\"],\"type\":\"string\"}},{\"description\":\"Search term for first input field\",\"example\":\"Shakespeare\",\"in\":\"path\",\"name\":\"searchTerm1\",\"required\":true,\"schema\":{\"type\":\"string\"}},{\"description\":\"Search term for second input field\",\"example\":\"14\",\"in\":\"path\",\"name\":\"searchTerm2\",\"required\":true,\"schema\":{\"type\":\"string\"}},{\"description\":\"Comma-separated list of output fields (author, title, lines, linecount, all)\",\"example\":\"lines\",\"in\":\"path\",\"name\":\"outputFields\",\"required\":true,\"schema\":{\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"items\":{\"type\":\"object\"},\"type\":\"array\"}}},\"description\":\"Successful response\"},\"404\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"reason\":{\"example\":\"Not found\",\"type\":\"string\"},\"status\":{\"example\":404,\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"No poems found\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/{inputField1},{inputField2}/{searchTerm1};{searchTerm2}/{outputFields}","rename":{"param":{"inputField1},{inputField2":"input_field1},{input_field2","outputFields":"output_field","searchTerm1};{searchTerm2":"search_term1};{search_term2"}},"segments":[{"lit":"{inputField1},{inputField2}"},{"lit":"{searchTerm1};{searchTerm2}"},{"var":"output_field"}],"select":{"exist":["input_field1","input_field2","output_field","search_term1","search_term2"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0}],"key$":"list"}},"relations":{"ancestors":[["{search_term1};{search_term2}"]]},"key$":"combined_search_with_field","name__orig":"combined_search_with_field","Name":"CombinedSearchWithField","name_":"combined_search_with_field","name-":"combined-search-with-field","NAME":"COMBINED_SEARCH_WITH_FIELD","index$":3}, {"active":true,"entity":"combined_search_with_field","key$":"BasicCombinedSearchWithFieldFlow","kind":"basic","name":"BasicCombinedSearchWithFieldFlow","param":{},"step":[{"active":true,"data":{},"input":{},"match":{"input_field1":"input_field101","input_field2":"input_field201","output_field":"output_field01","search_term1":"search_term101","search_term2":"search_term201"},"op":"list","spec":[],"valid":[{"apply":"ItemExists","def":{"ref":"combined_search_with_field_ref01"}}],"index$":0}]}, 'CombinedSearchWithField')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let combined_search_with_field_ref01_data = Object.values(setup.data.existing.combined_search_with_field)[0] as any

    // LIST
    const combined_search_with_field_ref01_ent = client.CombinedSearchWithField()
    const combined_search_with_field_ref01_match: any = {}
    combined_search_with_field_ref01_match['input_field1'] = setup.idmap['input_field101']
    combined_search_with_field_ref01_match['input_field2'] = setup.idmap['input_field201']
    combined_search_with_field_ref01_match['output_field'] = setup.idmap['output_field01']
    combined_search_with_field_ref01_match['search_term1'] = setup.idmap['search_term101']
    combined_search_with_field_ref01_match['search_term2'] = setup.idmap['search_term201']

    const combined_search_with_field_ref01_list = (await combined_search_with_field_ref01_ent.list(combined_search_with_field_ref01_match)).map((e: any) => e.data())


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/combined_search_with_field/CombinedSearchWithFieldTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = PoetrydbSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['combined_search_with_field01','combined_search_with_field02','combined_search_with_field03','{search_term1};{search_term2}01','{search_term1};{search_term2}02','{search_term1};{search_term2}03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'POETRYDB_TEST_COMBINED_SEARCH_WITH_FIELD_ENTID': idmap,
    'POETRYDB_TEST_LIVE': 'FALSE',
    'POETRYDB_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['POETRYDB_TEST_COMBINED_SEARCH_WITH_FIELD_ENTID']

  const live = 'TRUE' === env.POETRYDB_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['POETRYDB_TEST_COMBINED_SEARCH_WITH_FIELD_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new PoetrydbSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.POETRYDB_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  

File: rb/test/combined_search_with_field_entity_test.rb
# CombinedSearchWithField entity test

require "minitest/autorun"
require "json"
require_relative "../Poetrydb_sdk"
require_relative "runner"

class CombinedSearchWithFieldEntityTest < Minitest::Test
  def test_create_instance
    testsdk = PoetrydbSDK.test(nil, nil)
    ent = testsdk.CombinedSearchWithField(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "combined_search_with_field" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = PoetrydbSDK.test(seed, nil)
    seen = base.CombinedSearchWithField(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = PoetrydbConfig.make_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = PoetrydbSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.CombinedSearchWithField(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = combined_search_with_field_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "combined_search_with_field." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set POETRYDB_TEST_COMBINED_SEARCH_WITH_FIELD_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    combined_search_with_field_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.combined_search_with_field")))
    combined_search_with_field_ref01_data = nil
    if combined_search_with_field_ref01_data_raw.length > 0
      combined_search_with_field_ref01_data = Helpers.to_map(combined_search_with_field_ref01_data_raw[0][1])
    end

    # LIST
    combined_search_with_field_ref01_ent = client.CombinedSearchWithField(nil)
    combined_search_with_field_ref01_match = {
      "input_field1" => setup[:idmap]["input_field101"],
      "input_field2" => setup[:idmap]["input_field201"],
      "output_field" => setup[:idmap]["output_field01"],
      "search_term1" => setup[:idmap]["search_term101"],
      "search_term2" => setup[:idmap]["search_term201"],
    }

    combined_search_with_field_ref01_list_result = combined_search_with_field_ref01_ent.list(combined_search_with_field_ref01_match, nil)
    assert combined_search_with_field_ref01_list_result.is_a?(Array)

  end
end

def combined_search_with_field_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "combined_search_with_field", "CombinedSearchWithFieldTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = PoetrydbSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["combined_search_with_field01", "combined_search_with_field02", "combined_search_with_field03", "input_field101", "input_field201", "output_field01", "search_term101", "search_term201"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["POETRYDB_TEST_COMBINED_SEARCH_WITH_FIELD_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "POETRYDB_TEST_COMBINED_SEARCH_WITH_FIELD_ENTID" => idmap,
    "POETRYDB_TEST_LIVE" => "FALSE",
    "POETRYDB_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["POETRYDB_TEST_COMBINED_SEARCH_WITH_FIELD_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["POETRYDB_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = PoetrydbSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["POETRYDB_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["POETRYDB_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
